_posts/2020-09-31-spring-component-scanning.md

```diff
@@ -0,0 +1,252 @@
+
+**Specifying `@Configuration` in the default package will tell Spring to scan all the classes in all the jars in the classpath.**
+
+## How to Use `@ComponentScan`
+
+`@ComponentScan` along with `@Configuration` annotation is used to tell Spring to scan classes that are annotated with stereotype annotation. The `@ComponentScan` annotation provides different attributes that we can modify to get desired scanning behavior.
+
+```java
+@Component
+public class BeanViewer {
+
+  private final Logger LOG = LoggerFactory.getLogger(getClass());
+
+  @EventListener
+  public void showBeansRegistered(ApplicationReadyEvent event) {
+    String[] beanNames = event.getApplicationContext().getBeanDefinitionNames();
+      for(String beanName: beanNames) {
+        LOG.info("{}", beanName);
+    }
+  }
+}
+```
+
+We'll be using `ApplicationContext`'s `getBeanDefinitionNames()` method throughout this article to check out the list of beans created under the application context.
+
+The above `BeanViewer` will print all the beans that are registered with the application context. This will help us to check whether our components are loaded properly or not.
+
+### Spring Boot's Implicit Auto Scanning
+
+As said earlier, Spring Boot does auto scanning for all the packages that fall under the parent package. Let's have a look at how it works:
+
+```
+|- io.reflectoring.componentstan (main package)
+   |- SpringComponentScanningApplication.java
+   |- User.java (@Component stereotype)
+   |- BeanViewer.java
+   |- ExplicitScan.java
+```
+
+We have created a `User` class with the `@Component` stereotype in our parent package `io.reflectoring.componentscan`. As said earlier, since these classes come under the parent package where we have our main method defined with `@SpringBootApplication` annotation, the component will be scanned by default.
+
+```
+OUTPUT
+...
+INFO 95832 --- [main] i.reflectoring.componentscan.BeanViewer  : beanViewer
+INFO 95832 --- [main] i.reflectoring.componentscan.BeanViewer  : explicitScan
+INFO 95832 --- [main] i.reflectoring.componentscan.BeanViewer  : users
+...
+```
+
+The above output shows the bean created for `BeanViewer`, `ExplicitScan`, and `Users` and printed by our `BeanViewer`. Since all these classes come under our parent package `io.reflectoring.componentscan` these are auto scanned by Spring.
+
+### Using `@ComponentScan` without any attributes
+
+If we have a package that does not come under our parent package, we can use `@ComponentScan` along with `@Configuration`. This will tell Spring to scan the components in the package where this `@Configuration` class is present and its sub-packages.
+
+```java
```

Comment:
That's three code examples in a row without any "story" in between. Please have a look at how I changed this to make it more readable embedded in a "story".

Reply:
Yeah got it @thombergs. I'll work on my explanations.
Sorry for making you to go through lot of modifications.